src/server/coprocessor/executor/index_scan.rs

```diff
@@ -0,0 +1,318 @@
+        }
+        Ok(Some(Row::new(handle, values)))
+    }
+}
+
+impl<'a> Executor for IndexScanExec<'a> {
+    fn next(&mut self) -> Result<Option<Row>> {
+        while self.cursor < self.key_ranges.len() {
+            let data = box_try!(self.get_row_from_range());
+            if data.is_none() {
+                self.scanner.set_seek_key(None);
+                self.cursor += 1;
+                continue;
+            }
+            return Ok(data);
+        }
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use super::super::scanner::test::{Data, TestStore, new_col_info};
+    use util::codec::mysql::types;
+    use util::codec::datum::{self, Datum};
+    use util::codec::number::NumberEncoder;
+    use util::collections::HashMap;
+    use std::i64;
+    use tipb::schema::ColumnInfo;
+    use storage::Statistics;
+    use protobuf::RepeatedField;
+
+    const TABLE_ID: i64 = 1;
+    const INDEX_ID: i64 = 1;
+    const KEY_NUMBER: usize = 10;
+
+    #[inline]
+    pub fn get_idx_range(table_id: i64, idx_id: i64, start: i64, end: i64) -> KeyRange {
+        let mut start_buf = Vec::with_capacity(8);
+        start_buf.encode_i64(start).unwrap();
+        let mut end_buf = Vec::with_capacity(8);
+        end_buf.encode_i64(end).unwrap();
+        let mut key_range = KeyRange::new();
+        key_range.set_start(table::encode_index_seek_key(table_id, idx_id, &start_buf));
+        key_range.set_end(table::encode_index_seek_key(table_id, idx_id, &end_buf));
+        key_range
+    }
+
+    pub fn prepare_index_data(key_number: usize, table_id: i64, index_id: i64) -> Data {
+        let cols = vec![new_col_info(1, types::LONG_LONG),
+                        new_col_info(2, types::VARCHAR),
+                        new_col_info(3, types::NEW_DECIMAL)];
+
+        let mut data = Vec::new();
+        let mut pk = Vec::new();
+        let mut pk_handle = 0 as i64;
+        let mut encode_data = Vec::new();
+
+        for handle in 0..key_number {
+            let indice = map![
+                2 => Datum::Bytes(b"abc".to_vec()),
+                3 => Datum::Dec(10.into())
+            ];
+            let mut encode_value = HashMap::default();
+            let mut v: Vec<_> = indice.iter()
+                .map(|(k, value)| {
+                    encode_value.insert(*k, datum::encode_key(&[value.clone()]).unwrap());
+                    value.clone()
+                })
+                .collect();
+            let h = Datum::I64(handle as i64);
+            v.push(h);
+            let encoded = datum::encode_key(&v).unwrap();
+            let idx_key = table::encode_index_seek_key(table_id, index_id, &encoded);
+            if pk.is_empty() {
+                pk = idx_key.clone();
+                pk_handle = handle as i64;
+            }
+            encode_data.push(encode_value);
+            data.push((idx_key, vec![0]));
+        }
+        Data {
+            kv_data: data,
+            pk: pk,
```

Comment:
The `Data` has multiple rows, but the `pk` is for a single row, why `pk` is a member of `Data`?